File: llmcalc/usage_normalization.py
"""Normalize provider usage objects into billable token subsets."""

from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any, cast

MAX_SAFE_INTEGER = 2**53 - 1

TEXT_INPUT_MESSAGE = (
    "llmcalc takes token counts, not text. Pass integer input/output token counts, "
    "or the provider's usage object (for example response.usage). llmcalc does not "
    "tokenize strings or message lists."
)

_INPUT_KEYS = ("input_tokens", "prompt_tokens", "inputTokens", "promptTokens")
_OUTPUT_KEYS = ("output_tokens", "completion_tokens", "outputTokens", "completionTokens")
_GEMINI_INPUT_KEYS = ("prompt_token_count", "promptTokenCount")
_GEMINI_OUTPUT_KEYS = (
    "candidates_token_count",
    "candidatesTokenCount",
    "response_token_count",
    "responseTokenCount",
)
_GEMINI_CACHE_KEYS = ("cached_content_token_count", "cachedContentTokenCount")
_GEMINI_REASONING_KEYS = ("thoughts_token_count", "thoughtsTokenCount")
_GEMINI_TOOL_USE_KEYS = ("tool_use_prompt_token_count", "toolUsePromptTokenCount")
_GEMINI_TOTAL_KEYS = ("total_token_count", "totalTokenCount")
_PROMPT_DETAIL_KEYS = ("prompt_tokens_details", "promptTokensDetails", "input_tokens_details")
_COMPLETION_DETAIL_KEYS = (
    "completion_tokens_details",
    "completionTokensDetails",
    "output_tokens_details",
    "candidates_tokens_details",
    "candidatesTokensDetails",
    "response_tokens_details",
    "responseTokensDetails",
)
_CACHE_READ_KEYS = ("cache_read_input_tokens", "cacheReadInputTokens")
_CACHE_CREATION_KEYS = (
    "cache_creation_input_tokens",
    "cacheCreationInputTokens",
    "cache_write_input_tokens",
    "cacheWriteInputTokens",
)
_CACHED_SUBSET_KEYS = ("cached_tokens", "cachedTokens")
_CACHE_WRITE_SUBSET_KEYS = ("cache_write_tokens", "cacheWriteTokens")
_REASONING_KEYS = (
    "reasoning_tokens",
    "reasoningTokens",
    "thinking_tokens",
    "thinkingTokens",
)
_CACHE_CREATION_DETAIL_KEYS = ("cache_creation", "cacheCreation")
_FIVE_MINUTE_CACHE_KEYS = ("ephemeral_5m_input_tokens", "ephemeral5mInputTokens")
_ONE_HOUR_CACHE_KEYS = ("ephemeral_1h_input_tokens", "ephemeral1hInputTokens")
_BEDROCK_CACHE_DETAIL_KEYS = ("cache_details", "cacheDetails")
_BEDROCK_DETAIL_TOKEN_KEYS = ("input_tokens", "inputTokens")
_SERVICE_TIER_KEYS = ("service_tier", "serviceTier", "traffic_type", "trafficType")
_STANDARD_SERVICE_TIERS = {
    "0",
    "default",
    "standard",
    "standard_only",
    "on_demand",
    "service_tier_unspecified",
    "unspecified",
    "traffic_type_unspecified",
}
@dataclass(frozen=True)
class NormalizedUsage:
    input_tokens: int
    output_tokens: int
    cached_tokens: int = 0
    cache_creation_tokens: int = 0
    reasoning_tokens: int = 0
    processing_mode: str | None = None
    cache_creation_tokens_1h: int = 0
    cache_creation_audio_tokens: int = 0
    cached_audio_tokens: int = 0
    cached_image_tokens: int = 0
    input_audio_tokens: int = 0
    output_audio_tokens: int = 0
    input_image_tokens: int = 0
    output_image_tokens: int = 0
    query_count: int = 0
    web_search_requests: int = 0
    maps_grounding_requests: int = 0


def validate_token_count(value: int, field_name: str) -> None:
    if isinstance(value, bool) or not isinstance(value, int):
        raise ValueError(f"{field_name} must be an integer, got {type(value).__name__}")
    if value < 0:
        raise ValueError(f"{field_name} must be non-negative")
    if value > MAX_SAFE_INTEGER:
        raise ValueError(f"{field_name} must not exceed {MAX_SAFE_INTEGER}")


def _coerce_token_value(value: Any, field_name: str) -> int | None:
    if value is None:
        return None
    if type(value) is int:
        validate_token_count(value, field_name)
        return value
    raise ValueError(f"{field_name} must be an integer, got {type(value).__name__}")


def _value_from_mapping(
    usage: Mapping[str, Any], key_options: tuple[str, ...], field_name: str
) -> int | None:
    for key in key_options:
        if key in usage:
            value = _coerce_token_value(usage[key], field_name)
            if value is not None:
                return value
    return None


def _attr_token(usage: Any, names: tuple[str, ...], field_name: str) -> int | None:
    for name in names:
        value = getattr(usage, name, None)
        if value is not None:
            return _coerce_token_value(value, field_name)
    return None


def _nested(usage: Any, container_keys: tuple[str, ...], names: tuple[str, ...]) -> int:
    for container_key in container_keys:
        container = (
            usage.get(container_key)
            if isinstance(usage, Mapping)
            else getattr(usage, container_key, None)
        )
        if container is None:
            continue
        value = (
            _value_from_mapping(container, names, container_key)
            if isinstance(container, Mapping)
            else _attr_token(container, names, container_key)
        )
        if value is not None:
            return value
    return 0


def _flat(usage: Any, names: tuple[str, ...], field_name: str) -> int | None:
    if isinstance(usage, Mapping):
        return _value_from_mapping(usage, names, field_name)
    return _attr_token(usage, names, field_name)


def _raw_value(value: object, names: tuple[str, ...]) -> object | None:
    for name in names:
        candidate = value.get(name) if isinstance(value, Mapping) else getattr(value, name, None)
        if candidate is not None:
            return cast(object, candidate)
    return None


def _processing_mode(usage: object) -> str | None:
    tier = _raw_value(usage, _SERVICE_TIER_KEYS)
    if tier is None:
        return None
    raw_tier = getattr(tier, "value", tier)
    normalized = str(raw_tier).strip().lower().replace("-", "_")
    if normalized in _STANDARD_SERVICE_TIERS or normalized in {
        "auto",
        "on_demand",
        "provisioned_throughput",
    }:
        return "standard"
    if normalized in {"priority", "fast", "on_demand_priority"}:
        return "priority"
    if normalized in {"flex", "on_demand_flex"}:
        return "flex"
    if normalized in {"batch", "batches"}:
        return "batch"
    raise ValueError(f"service tier {raw_tier!s} is not supported")


def _modality_counts(details: object | None, field_name: str) -> tuple[int, int]:
    if details is None:
        return 0, 0
    if isinstance(details, (list, tuple)):
        audio = 0
        image = 0
        for detail in details:
            modality = _raw_value(detail, ("modality",))
            tokens = _flat(detail, ("token_count", "tokenCount"), field_name)
            if modality is None or tokens is None:
                continue
            normalized = str(getattr(modality, "value", modality)).strip().lower()
            if normalized == "audio":
                audio += tokens
            elif normalized in {"image", "document"}:
                image += tokens
            elif normalized not in {"text", "modality_unspecified"} and tokens > 0:
                raise ValueError(f"unsupported token modality: {normalized}")
        validate_token_count(audio, f"{field_name}.audio_tokens")
        validate_token_count(image, f"{field_name}.image_tokens")
        return audio, image

    audio = _flat(details, ("audio_tokens", "audioTokens"), f"{field_name}.audio_tokens") or 0
    image = _flat(details, ("image_tokens", "imageTokens"), f"{field_name}.image_tokens") or 0
    video = _flat(details, ("video_tokens", "videoTokens"), f"{field_name}.video_tokens") or 0
    if video > 0:
        raise ValueError("video token pricing is not supported")
    return audio, image


def _first_details(usage: object, keys: tuple[str, ...]) -> object | None:
    for key in keys:
        details = _raw_value(usage, (key,))
        if details is not None:
            return details
    return None


def _cache_duration_counts(usage: object) -> tuple[int, int]:
    five_minute = _nested(usage, _CACHE_CREATION_DETAIL_KEYS, _FIVE_MINUTE_CACHE_KEYS)
    one_hour = _nested(usage, _CACHE_CREATION_DETAIL_KEYS, _ONE_HOUR_CACHE_KEYS)
    detail_total = five_minute + one_hour
    details = _raw_value(usage, _BEDROCK_CACHE_DETAIL_KEYS)
    if isinstance(details, (list, tuple)):
        for detail in details:
            tokens = _flat(detail, _BEDROCK_DETAIL_TOKEN_KEYS, "cache_details.input_tokens")
            if tokens is None:
                continue
            detail_total += tokens
            if str(_raw_value(detail, ("ttl",))).strip().lower() == "1h":
                one_hour += tokens
    validate_token_count(detail_total, "cache_details.input_tokens")
    validate_token_count(one_hour, "cache_creation_tokens_1h")
    return detail_total, one_hour


def _get_base_usage_tokens(usage: object) -> tuple[int, int, int, int, int]:
    """Normalize provider usage to input, output, cache and reasoning counts."""
    if isinstance(usage, (str, list, tuple)):
        raise ValueError(TEXT_INPUT_MESSAGE)

    if isinstance(usage, Mapping) and "messages" in usage:
        raise ValueError(TEXT_INPUT_MESSAGE)

    gemini_input = _flat(usage, _GEMINI_INPUT_KEYS, "prompt_token_count")
    if gemini_input is not None:
        gemini_output = _flat(usage, _GEMINI_OUTPUT_KEYS, "candidates_token_count") or 0
        cached = _flat(usage, _GEMINI_CACHE_KEYS, "cached_content_token_count") or 0
        reasoning = _flat(usage, _GEMINI_REASONING_KEYS, "thoughts_token_count") or 0
        tool_use = _flat(usage, _GEMINI_TOOL_USE_KEYS, "tool_use_prompt_token_count") or 0
        total = _flat(usage, _GEMINI_TOTAL_KEYS, "total_token_count")
        validate_token_count(gemini_input, "prompt_token_count")
        validate_token_count(gemini_output, "candidates_token_count")
        validate_token_count(cached, "cached_content_token_count")
        validate_token_count(reasoning, "thoughts_token_count")
        validate_token_count(tool_use, "tool_use_prompt_token_count")
        if total is not None:
            validate_token_count(total, "total_token_count")
        if cached > gemini_input:
            raise ValueError("cached_content_token_count must not exceed prompt_token_count")

        output_total = gemini_output + reasoning
        total_without_tool = gemini_input + output_total
        validate_token_count(output_total, "output_tokens")
        validate_token_count(total_without_tool, "total_token_count")

        if tool_use == 0:
            if total is not None and total != total_without_tool:
                raise ValueError("total_token_count is inconsistent with Gemini token details")
            return gemini_input, output_total, cached, 0, reasoning

        if total is None:
            raise ValueError(
                "total_token_count is required to determine whether "
                "tool_use_prompt_token_count is additive"
            )
        if total == total_without_tool:
            return gemini_input, output_total, cached, 0, reasoning
        if total == total_without_tool + tool_use:
            input_total = gemini_input + tool_use
            validate_token_count(input_total, "input_tokens")
            return input_total, output_total, cached, 0, reasoning
        raise ValueError("total_token_count is inconsistent with Gemini token details")

    input_tokens = _flat(usage, _INPUT_KEYS, "input_tokens")
    output_tokens = _flat(usage, _OUTPUT_KEYS, "output_tokens")

    if input_tokens is None or output_tokens is None:
        raise ValueError("usage must provide input/prompt tokens and output/completion tokens")

    validate_token_count(input_tokens, "input_tokens")
    validate_token_count(output_tokens, "output_tokens")
    additive_read = _flat(usage, _CACHE_READ_KEYS, "cache_read_input_tokens")
    additive_creation = _flat(usage, _CACHE_CREATION_KEYS, "cache_creation_input_tokens")

    if additive_read is not None or additive_creation is not None:
        cached = additive_read or 0
        creation = additive_creation or 0
        validate_token_count(cached, "cache_read_input_tokens")
        validate_token_count(creation, "cache_creation_input_tokens")
        input_tokens += cached + creation
        validate_token_count(input_tokens, "input_tokens")
    else:
        cached = _nested(usage, _PROMPT_DETAIL_KEYS, _CACHED_SUBSET_KEYS)
        creation = _nested(usage, _PROMPT_DETAIL_KEYS, _CACHE_WRITE_SUBSET_KEYS)

    reasoning = _nested(usage, _COMPLETION_DETAIL_KEYS, _REASONING_KEYS)
    return input_tokens, output_tokens, cached, creation, reasoning


def _usage_counter(usage: object, names: tuple[str, ...], field_name: str) -> int:
    direct = _flat(usage, names, field_name)
    if direct is not None:
        return direct
    prompt_details = _first_details(usage, _PROMPT_DETAIL_KEYS)
    if prompt_details is not None and not isinstance(prompt_details, (list, tuple)):
        nested = _flat(prompt_details, names, field_name)
        if nested is not None:
            return nested
    return 0


def normalize_usage(usage: object) -> NormalizedUsage:
    input_tokens, output_tokens, cached, creation, reasoning = _get_base_usage_tokens(usage)
    prompt_details = _first_details(usage, _PROMPT_DETAIL_KEYS)
    completion_details = _first_details(usage, _COMPLETION_DETAIL_KEYS)
    input_audio, input_image = _modality_counts(prompt_details, "prompt_tokens_details")
    output_audio, output_image = _modality_counts(
        completion_details, "completion_tokens_details"
    )

    cache_details = _first_details(
        usage,
        ("cache_tokens_details", "cacheTokensDetails"),
    )
    if cache_details is None and prompt_details is not None and not isinstance(
        prompt_details, (list, tuple)
    ):
        cache_details = _raw_value(
            prompt_details,
            ("cached_tokens_details", "cachedTokensDetails"),
        )
    cached_audio, cached_image = _modality_counts(cache_details, "cache_tokens_details")

    detail_total, one_hour = _cache_duration_counts(usage)
    if detail_total > 0:
        if creation > 0 and detail_total != creation:
            raise ValueError(
                "cache duration token details must sum to cache_creation_input_tokens"
            )
        if creation == 0:
            creation = detail_total
            input_tokens += detail_total
            validate_token_count(input_tokens, "input_tokens")
    if one_hour > creation:
        raise ValueError("1-hour cache tokens must not exceed cache creation tokens")

    server_tool_use = _raw_value(usage, ("server_tool_use", "serverToolUse"))
    web_search_requests = _usage_counter(
        usage,
        ("web_search_requests", "webSearchRequests"),
        "web_search_requests",
    )
    if web_search_requests == 0 and server_tool_use is not None:
        web_search_requests = (
            _flat(
                server_tool_use,
                ("web_search_requests", "webSearchRequests"),
                "server_tool_use.web_search_requests",
            )
            or 0
        )
    server_side_tool_use = _raw_value(
        usage,
        ("server_side_tool_usage_details", "serverSideToolUsageDetails"),
    )
    if web_search_requests == 0 and server_side_tool_use is not None:
        web_search_requests = (
            _flat(
                server_side_tool_use,
                ("web_search_calls", "webSearchCalls"),
                "server_side_tool_usage_details.web_search_calls",
            )
            or 0
        )

    normalized = NormalizedUsage(
        input_tokens=input_tokens,
        output_tokens=output_tokens,
        cached_tokens=cached,
        cache_creation_tokens=creation,
        reasoning_tokens=reasoning,
        processing_mode=_processing_mode(usage),
        cache_creation_tokens_1h=one_hour,
        cache_creation_audio_tokens=_usage_counter(
            usage,
            ("cache_creation_audio_tokens", "cacheCreationAudioTokens"),
            "cache_creation_audio_tokens",
        ),
        cached_audio_tokens=cached_audio,
        cached_image_tokens=cached_image,
        input_audio_tokens=input_audio,
        output_audio_tokens=output_audio,
        input_image_tokens=input_image,
        output_image_tokens=output_image,
        query_count=_usage_counter(
            usage,
            ("query_count", "queryCount", "search_units", "searchUnits"),
            "query_count",
        ),
        web_search_requests=web_search_requests,
        maps_grounding_requests=_usage_counter(
            usage,
            ("google_maps_grounding_requests", "googleMapsGroundingRequests"),
            "google_maps_grounding_requests",
        ),
    )
    for field_name, value in normalized.__dict__.items():
        if field_name != "processing_mode":
            validate_token_count(value, field_name)
    return normalized


def get_usage_tokens(usage: object) -> tuple[int, int, int, int, int]:
    normalized = normalize_usage(usage)
    return (
        normalized.input_tokens,
        normalized.output_tokens,
        normalized.cached_tokens,
        normalized.cache_creation_tokens,
        normalized.reasoning_tokens,
    )

File: js/src/usage-normalization.ts
export const TEXT_INPUT_MESSAGE =
  "llmcalc takes token counts, not text. Pass integer input/output token counts, " +
  "or the provider's usage object (for example response.usage). llmcalc does not " +
  "tokenize strings or message lists.";

const INPUT_KEYS = ["input_tokens", "prompt_tokens", "inputTokens", "promptTokens"] as const;
const OUTPUT_KEYS = [
  "output_tokens",
  "completion_tokens",
  "outputTokens",
  "completionTokens"
] as const;
const GEMINI_INPUT_KEYS = ["prompt_token_count", "promptTokenCount"] as const;
const GEMINI_OUTPUT_KEYS = [
  "candidates_token_count",
  "candidatesTokenCount",
  "response_token_count",
  "responseTokenCount"
] as const;
const GEMINI_CACHE_KEYS = ["cached_content_token_count", "cachedContentTokenCount"] as const;
const GEMINI_REASONING_KEYS = ["thoughts_token_count", "thoughtsTokenCount"] as const;
const GEMINI_TOOL_USE_KEYS = [
  "tool_use_prompt_token_count",
  "toolUsePromptTokenCount"
] as const;
const GEMINI_TOTAL_KEYS = ["total_token_count", "totalTokenCount"] as const;
const PROMPT_DETAIL_KEYS = [
  "prompt_tokens_details",
  "promptTokensDetails",
  "input_tokens_details"
] as const;
const COMPLETION_DETAIL_KEYS = [
  "completion_tokens_details",
  "completionTokensDetails",
  "output_tokens_details",
  "candidates_tokens_details",
  "candidatesTokensDetails",
  "response_tokens_details",
  "responseTokensDetails"
] as const;
const CACHE_READ_KEYS = ["cache_read_input_tokens", "cacheReadInputTokens"] as const;
const CACHE_CREATION_KEYS = [
  "cache_creation_input_tokens",
  "cacheCreationInputTokens",
  "cache_write_input_tokens",
  "cacheWriteInputTokens"
] as const;
const CACHED_SUBSET_KEYS = ["cached_tokens", "cachedTokens"] as const;
const CACHE_WRITE_SUBSET_KEYS = ["cache_write_tokens", "cacheWriteTokens"] as const;
const REASONING_KEYS = [
  "reasoning_tokens",
  "reasoningTokens",
  "thinking_tokens",
  "thinkingTokens"
] as const;
const CACHE_CREATION_DETAIL_KEYS = ["cache_creation", "cacheCreation"] as const;
const FIVE_MINUTE_CACHE_KEYS = ["ephemeral_5m_input_tokens", "ephemeral5mInputTokens"] as const;
const ONE_HOUR_CACHE_KEYS = ["ephemeral_1h_input_tokens", "ephemeral1hInputTokens"] as const;
const BEDROCK_CACHE_DETAIL_KEYS = ["cache_details", "cacheDetails"] as const;
const BEDROCK_DETAIL_TOKEN_KEYS = ["input_tokens", "inputTokens"] as const;
const SERVICE_TIER_KEYS = ["service_tier", "serviceTier", "traffic_type", "trafficType"] as const;
const STANDARD_SERVICE_TIERS = new Set([
  "0",
  "default",
  "standard",
  "standard_only",
  "on_demand",
  "service_tier_unspecified",
  "unspecified",
  "traffic_type_unspecified"
]);
export interface NormalizedUsage {
  inputTokens: number;
  outputTokens: number;
  cachedTokens: number;
  cacheCreationTokens: number;
  reasoningTokens: number;
  processingMode: string | null;
  cacheCreationTokens1h: number;
  cacheCreationAudioTokens: number;
  cachedAudioTokens: number;
  cachedImageTokens: number;
  inputAudioTokens: number;
  outputAudioTokens: number;
  inputImageTokens: number;
  outputImageTokens: number;
  queryCount: number;
  webSearchRequests: number;
  mapsGroundingRequests: number;
}

export function validateTokenCount(value: number, fieldName: string): void {
  if (typeof value !== "number" || !Number.isSafeInteger(value)) {
    throw new Error(`${fieldName} must be an integer, got ${typeof value}`);
  }
  if (value < 0) {
    throw new Error(`${fieldName} must be non-negative`);
  }
}

function coerceTokenValue(value: unknown, fieldName: string): number {
  if (typeof value !== "number" || !Number.isSafeInteger(value)) {
    throw new Error(`${fieldName} must be an integer, got ${typeof value}`);
  }
  return value;
}

function valueFromMapping(
  usage: Record<string, unknown>,
  keyOptions: readonly string[],
  fieldName: string
): number | null {
  for (const key of keyOptions) {
    if (key in usage && usage[key] !== null && usage[key] !== undefined) {
      return coerceTokenValue(usage[key], fieldName);
    }
  }
  return null;
}

function nestedToken(
  usage: Record<string, unknown>,
  containerKeys: readonly string[],
  names: readonly string[]
): number {
  for (const containerKey of containerKeys) {
    const container = usage[containerKey];
    if (typeof container !== "object" || container === null) {
      continue;
    }
    const value = valueFromMapping(container as Record<string, unknown>, names, containerKey);
    if (value !== null) {
      return value;
    }
  }
  return 0;
}

function rawValue(usage: Record<string, unknown>, names: readonly string[]): unknown {
  for (const name of names) {
    const value = usage[name];
    if (value !== null && value !== undefined) {
      return value;
    }
  }
  return null;
}

export function getUsageTokens(usage: unknown): [number, number, number, number, number] {
  if (typeof usage === "string" || Array.isArray(usage)) {
    throw new Error(TEXT_INPUT_MESSAGE);
  }
  if (typeof usage !== "object" || usage === null) {
    throw new Error("usage must provide input/prompt tokens and output/completion tokens");
  }

  const dictUsage = usage as Record<string, unknown>;
  if ("messages" in dictUsage) {
    throw new Error(TEXT_INPUT_MESSAGE);
  }

  const geminiInput = valueFromMapping(dictUsage, GEMINI_INPUT_KEYS, "promptTokenCount");
  if (geminiInput !== null) {
    const geminiOutput = valueFromMapping(
      dictUsage,
      GEMINI_OUTPUT_KEYS,
      "candidatesTokenCount"
    ) ?? 0;
    const cached = valueFromMapping(dictUsage, GEMINI_CACHE_KEYS, "cachedContentTokenCount") ?? 0;
    const reasoning = valueFromMapping(dictUsage, GEMINI_REASONING_KEYS, "thoughtsTokenCount") ?? 0;
    const toolUse = valueFromMapping(
      dictUsage,
      GEMINI_TOOL_USE_KEYS,
      "toolUsePromptTokenCount"
    ) ?? 0;
    const total = valueFromMapping(dictUsage, GEMINI_TOTAL_KEYS, "totalTokenCount");
    validateTokenCount(geminiInput, "promptTokenCount");
    validateTokenCount(geminiOutput, "candidatesTokenCount");
    validateTokenCount(cached, "cachedContentTokenCount");
    validateTokenCount(reasoning, "thoughtsTokenCount");
    validateTokenCount(toolUse, "toolUsePromptTokenCount");
    if (total !== null) {
      validateTokenCount(total, "totalTokenCount");
    }
    if (cached > geminiInput) {
      throw new Error("cachedContentTokenCount must not exceed promptTokenCount");
    }

    const outputTotal = geminiOutput + reasoning;
    const totalWithoutTool = geminiInput + outputTotal;
    validateTokenCount(outputTotal, "outputTokens");
    validateTokenCount(totalWithoutTool, "totalTokenCount");

    if (toolUse === 0) {
      if (total !== null && total !== totalWithoutTool) {
        throw new Error("totalTokenCount is inconsistent with Gemini token details");
      }
      return [geminiInput, outputTotal, cached, 0, reasoning];
    }
    if (total === null) {
      throw new Error(
        "totalTokenCount is required to determine whether toolUsePromptTokenCount is additive"
      );
    }
    if (total === totalWithoutTool) {
      return [geminiInput, outputTotal, cached, 0, reasoning];
    }
    if (total === totalWithoutTool + toolUse) {
      const inputTotal = geminiInput + toolUse;
      validateTokenCount(inputTotal, "inputTokens");
      return [inputTotal, outputTotal, cached, 0, reasoning];
    }
    throw new Error("totalTokenCount is inconsistent with Gemini token details");
  }

  let inputTokens = valueFromMapping(dictUsage, INPUT_KEYS, "inputTokens");
  const outputTokens = valueFromMapping(dictUsage, OUTPUT_KEYS, "outputTokens");
  if (inputTokens === null || outputTokens === null) {
    throw new Error("usage must provide input/prompt tokens and output/completion tokens");
  }

  validateTokenCount(inputTokens, "inputTokens");
  validateTokenCount(outputTokens, "outputTokens");
  const additiveRead = valueFromMapping(dictUsage, CACHE_READ_KEYS, "cacheReadInputTokens");
  const additiveCreation = valueFromMapping(
    dictUsage,
    CACHE_CREATION_KEYS,
    "cacheCreationInputTokens"
  );

  let cached: number;
  let creation: number;
  if (additiveRead !== null || additiveCreation !== null) {
    cached = additiveRead ?? 0;
    creation = additiveCreation ?? 0;
    validateTokenCount(cached, "cacheReadInputTokens");
    validateTokenCount(creation, "cacheCreationInputTokens");
    inputTokens += cached + creation;
    validateTokenCount(inputTokens, "inputTokens");
  } else {
    cached = nestedToken(dictUsage, PROMPT_DETAIL_KEYS, CACHED_SUBSET_KEYS);
    creation = nestedToken(dictUsage, PROMPT_DETAIL_KEYS, CACHE_WRITE_SUBSET_KEYS);
  }

  const reasoning = nestedToken(dictUsage, COMPLETION_DETAIL_KEYS, REASONING_KEYS);
  return [inputTokens, outputTokens, cached, creation, reasoning];
}

function processingMode(usage: Record<string, unknown>): string | null {
  const tier = rawValue(usage, SERVICE_TIER_KEYS);
  if (tier === null) {
    return null;
  }
  const normalized = String(tier).trim().toLowerCase().replaceAll("-", "_");
  if (
    STANDARD_SERVICE_TIERS.has(normalized) ||
    normalized === "auto" ||
    normalized === "provisioned_throughput"
  ) {
    return "standard";
  }
  if (normalized === "priority" || normalized === "fast" || normalized === "on_demand_priority") {
    return "priority";
  }
  if (normalized === "flex" || normalized === "on_demand_flex") {
    return "flex";
  }
  if (normalized === "batch" || normalized === "batches") {
    return "batch";
  }
  throw new Error(`service tier ${String(tier)} is not supported`);
}

function firstDetails(
  usage: Record<string, unknown>,
  keys: readonly string[]
): unknown {
  for (const key of keys) {
    const details = usage[key];
    if (details !== null && details !== undefined) {
      return details;
    }
  }
  return null;
}

function modalityCounts(details: unknown, fieldName: string): [number, number] {
  if (details === null || details === undefined) {
    return [0, 0];
  }
  if (Array.isArray(details)) {
    let audio = 0;
    let image = 0;
    for (const item of details) {
      if (typeof item !== "object" || item === null || Array.isArray(item)) {
        continue;
      }
      const record = item as Record<string, unknown>;
      const modality = rawValue(record, ["modality"]);
      const tokens = valueFromMapping(record, ["token_count", "tokenCount"], fieldName);
      if (modality === null || tokens === null) {
        continue;
      }
      const normalized = String(modality).trim().toLowerCase();
      if (normalized === "audio") {
        audio += tokens;
      } else if (normalized === "image" || normalized === "document") {
        image += tokens;
      } else if (normalized !== "text" && normalized !== "modality_unspecified" && tokens > 0) {
        throw new Error(`unsupported token modality: ${normalized}`);
      }
    }
    validateTokenCount(audio, `${fieldName}.audioTokens`);
    validateTokenCount(image, `${fieldName}.imageTokens`);
    return [audio, image];
  }
  if (typeof details !== "object") {
    return [0, 0];
  }
  const record = details as Record<string, unknown>;
  const audio = valueFromMapping(record, ["audio_tokens", "audioTokens"], `${fieldName}.audioTokens`) ?? 0;
  const image = valueFromMapping(record, ["image_tokens", "imageTokens"], `${fieldName}.imageTokens`) ?? 0;
  const video = valueFromMapping(record, ["video_tokens", "videoTokens"], `${fieldName}.videoTokens`) ?? 0;
  if (video > 0) {
    throw new Error("video token pricing is not supported");
  }
  return [audio, image];
}

function cacheDurationCounts(usage: Record<string, unknown>): [number, number] {
  const fiveMinute = nestedToken(usage, CACHE_CREATION_DETAIL_KEYS, FIVE_MINUTE_CACHE_KEYS);
  let oneHour = nestedToken(usage, CACHE_CREATION_DETAIL_KEYS, ONE_HOUR_CACHE_KEYS);
  let detailTotal = fiveMinute + oneHour;
  const details = rawValue(usage, BEDROCK_CACHE_DETAIL_KEYS);
  if (Array.isArray(details)) {
    for (const item of details) {
      if (typeof item !== "object" || item === null || Array.isArray(item)) {
        continue;
      }
      const record = item as Record<string, unknown>;
      const tokens = valueFromMapping(record, BEDROCK_DETAIL_TOKEN_KEYS, "cacheDetails.inputTokens");
      if (tokens === null) {
        continue;
      }
      detailTotal += tokens;
      if (String(rawValue(record, ["ttl"])).trim().toLowerCase() === "1h") {
        oneHour += tokens;
      }
    }
  }
  validateTokenCount(detailTotal, "cacheDetails.inputTokens");
  validateTokenCount(oneHour, "cacheCreationTokens1h");
  return [detailTotal, oneHour];
}

function usageCounter(
  usage: Record<string, unknown>,
  names: readonly string[],
  fieldName: string
): number {
  const direct = valueFromMapping(usage, names, fieldName);
  if (direct !== null) {
    return direct;
  }
  const promptDetails = firstDetails(usage, PROMPT_DETAIL_KEYS);
  if (typeof promptDetails === "object" && promptDetails !== null && !Array.isArray(promptDetails)) {
    return valueFromMapping(promptDetails as Record<string, unknown>, names, fieldName) ?? 0;
  }
  return 0;
}

export function normalizeUsage(usage: unknown): NormalizedUsage {
  const [baseInput, outputTokens, cachedTokens, baseCreation, reasoningTokens] =
    getUsageTokens(usage);
  const dictUsage = usage as Record<string, unknown>;
  const promptDetails = firstDetails(dictUsage, PROMPT_DETAIL_KEYS);
  const completionDetails = firstDetails(dictUsage, COMPLETION_DETAIL_KEYS);
  const [inputAudioTokens, inputImageTokens] = modalityCounts(
    promptDetails,
    "promptTokensDetails"
  );
  const [outputAudioTokens, outputImageTokens] = modalityCounts(
    completionDetails,
    "completionTokensDetails"
  );

  let cacheDetails = firstDetails(dictUsage, ["cache_tokens_details", "cacheTokensDetails"]);
  if (
    cacheDetails === null &&
    typeof promptDetails === "object" &&
    promptDetails !== null &&
    !Array.isArray(promptDetails)
  ) {
    cacheDetails = rawValue(
      promptDetails as Record<string, unknown>,
      ["cached_tokens_details", "cachedTokensDetails"]
    );
  }
  const [cachedAudioTokens, cachedImageTokens] = modalityCounts(
    cacheDetails,
    "cacheTokensDetails"
  );

  const [detailTotal, cacheCreationTokens1h] = cacheDurationCounts(dictUsage);
  let inputTokens = baseInput;
  let cacheCreationTokens = baseCreation;
  if (detailTotal > 0) {
    if (cacheCreationTokens > 0 && detailTotal !== cacheCreationTokens) {
      throw new Error("cache duration token details must sum to cacheCreationInputTokens");
    }
    if (cacheCreationTokens === 0) {
      cacheCreationTokens = detailTotal;
      inputTokens += detailTotal;
      validateTokenCount(inputTokens, "inputTokens");
    }
  }
  if (cacheCreationTokens1h > cacheCreationTokens) {
    throw new Error("1-hour cache tokens must not exceed cache creation tokens");
  }

  let webSearchRequests = usageCounter(
    dictUsage,
    ["web_search_requests", "webSearchRequests"],
    "webSearchRequests"
  );
  const serverToolUse = rawValue(dictUsage, ["server_tool_use", "serverToolUse"]);
  if (
    webSearchRequests === 0 &&
    typeof serverToolUse === "object" &&
    serverToolUse !== null &&
    !Array.isArray(serverToolUse)
  ) {
    webSearchRequests = valueFromMapping(
      serverToolUse as Record<string, unknown>,
      ["web_search_requests", "webSearchRequests"],
      "serverToolUse.webSearchRequests"
    ) ?? 0;
  }
  const serverSideToolUse = rawValue(dictUsage, [
    "server_side_tool_usage_details",
    "serverSideToolUsageDetails"
  ]);
  if (
    webSearchRequests === 0 &&
    typeof serverSideToolUse === "object" &&
    serverSideToolUse !== null &&
    !Array.isArray(serverSideToolUse)
  ) {
    webSearchRequests = valueFromMapping(
      serverSideToolUse as Record<string, unknown>,
      ["web_search_calls", "webSearchCalls"],
      "serverSideToolUsageDetails.webSearchCalls"
    ) ?? 0;
  }

  return {
    inputTokens,
    outputTokens,
    cachedTokens,
    cacheCreationTokens,
    reasoningTokens,
    processingMode: processingMode(dictUsage),
    cacheCreationTokens1h,
    cacheCreationAudioTokens: usageCounter(
      dictUsage,
      ["cache_creation_audio_tokens", "cacheCreationAudioTokens"],
      "cacheCreationAudioTokens"
    ),
    cachedAudioTokens,
    cachedImageTokens,
    inputAudioTokens,
    outputAudioTokens,
    inputImageTokens,
    outputImageTokens,
    queryCount: usageCounter(
      dictUsage,
      ["query_count", "queryCount", "search_units", "searchUnits"],
      "queryCount"
    ),
    webSearchRequests,
    mapsGroundingRequests: usageCounter(
      dictUsage,
      ["google_maps_grounding_requests", "googleMapsGroundingRequests"],
      "googleMapsGroundingRequests"
    )
  };
}
